### ISSUE TITLE
Missing Swing Die after setting its value

### ISSUE THREAD

**Issue by TheOrgg** [buttonmen-dev/buttonmen#2749]
http://www.buttonweavers.com/ui/game.html?game=76655

In a Hit List game, I set my X to 4, but it didn't appear in my recipe.

**Comment by blackshadowshade** [comment]
It looks like the die has disappeared completely!

That's definitely a bug.

Can you tell us if the same happens next round too?

**Comment by TheOrgg** [comment]
Will do.  Is there something in the background that can tell you what happened?

**Comment by jl8e** [comment]
Just speculating here:

(2021-08-14 22:20:14) | ElihuRoot set swing values: X=16
(2021-08-14 22:20:13) | TheOrgg set swing values: X=4

Is a race condition possible?

**Comment by jl8e** [comment]
Elaborating a bit more:

Without going diving into the guts of the engine, I think there are two potential race types. One is easy to test, but unfortunately very unlikely to be the cause. The other is a pain to test.

The easy one:

TheOrgg accepts the game, goes into swing die selection. Game state is currently no swings set.
ElihuRoot goes to the game, goes into swing die selection. Game state is currently no swings set.
TheOrgg submits their swing die choice. Game state updates.
ElihuRoot submits their swing die choice. Game state updates, but there's something being set incorrectly because the request was made by a client that thinks there are no swing dice set.

As I said, I think this is very unlikely. I haven't checked, but I'd assume that the client has no idea whether the opponent has set their swings or not, and the request is identical in either case.

The hard one:

I honestly don't know enough about the guts of how the code interacts with the DB, or how the DB behaves (I believe the engine is single-threaded, but is the DB?) to speculate on the mechanics at this level, but it seems not impossible that some of the various DB activities from TheOrgg's request were still pending when ElihuRoot's request was processed.

**Comment by cgolubi1** [comment]
It's definitely possible.  Here's the apache log:

```
X.X.X.X - - [15/Aug/2021:02:19:51 +0000] "POST /api/responder HTTP/1.1" 200 3219 "http://www.buttonweavers.com/ui/game.html?game=76655" "<BROWSER>" BMUser=<ElihuRoot> BMAPI=loadGameData BMStatus=ok
Y.Y.Y.Y - - [15/Aug/2021:02:20:06 +0000] "POST /api/responder HTTP/1.1" 200 3222 "http://www.buttonweavers.com/ui/game.html?game=76655" "<BROWSER>" BMUser=<TheOrgg> BMAPI=loadGameData BMStatus=ok
Y.Y.Y.Y - - [15/Aug/2021:02:20:13 +0000] "POST /api/responder HTTP/1.1" 200 66 "http://www.buttonweavers.com/ui/game.html?game=76655" "<BROWSER>" BMUser=<TheOrgg> BMAPI=submitDieValues BMStatus=ok
Y.Y.Y.Y - - [15/Aug/2021:02:20:14 +0000] "POST /api/responder HTTP/1.1" 200 3425 "http://www.buttonweavers.com/ui/game.html?game=76655" "<BROWSER>" BMUser=<TheOrgg> BMAPI=loadGameData BMStatus=ok
X.X.X.X - - [15/Aug/2021:02:20:13 +0000] "POST /api/responder HTTP/1.1" 200 66 "http://www.buttonweavers.com/ui/game.html?game=76655" "<BROWSER>" BMUser=<ElihuRoot> BMAPI=submitDieValues BMStatus=ok
```

**Comment by cgolubi1** [comment]
Yeah, so there are two pieces of this which might be relevant:
* ElihuRoot loaded game data first, but TheOrgg submitted die values first --- i just did a test on staging, and in isolation this does not break everything (and really if it did, this would have popped up by now)
* More likely to be relevant: TheOrgg's `submitDieValues` shows up first in the apache log, but with a later timestamp.  That implies that ElihuRoot's was actually started first, but completed second.

tl;dr: i'm on board with jl8e's theory that the timing is very likely to be relevant.  I'll investigate further with that as a starting point.

**Comment by cgolubi1** [comment]
Okay, so i looked at `submit_die_values()`, and it definitely seems to me like *if* there's a race possible between distinct database changes performed by `save_game()`, nothing in `submit_die_values()` would prevent that from landing us in a weird state in a case like this one.

See this conversation between me and shadowshade about a different weird state i hit while stress-testing customrecipes, which we also thought was potentially a `save_game()` race or partial failure: https://github.com/buttonmen-dev/buttonmen/pull/2389#issuecomment-825620036 and https://github.com/buttonmen-dev/buttonmen/pull/2389#issuecomment-825657777.  In particular, i wonder if we could just wrap all of `save_game()`'s database modifications in a transaction, and this problem would go away (i.e. lead to a visible error without corrupting the database) without us having to understand precisely what happened.

**Comment by cgolubi1** [comment]
I'm going to close this, on the theory that #2750 addressed it as well as we know how based on our understanding of what happened.  Feel free to disagree and reopen!
